doc/src/devdocs/types.md

```diff
@@ -315,7 +315,129 @@ Ptr{Void} @0x00007f5998a04370
 
 so `Tuple == Tuple{Vararg{Any}}` is indeed the primary type.
 
-## Introduction to the internal machinery: `jltypes.c`
+## Diagonal types
+
+Consider the type `Tuple{T,T} where T`.
+A method with this signature would look like:
+
+```julia
+f(x::T, y::T) where T = ...
+```
+
+According to the usual interpretation of a `UnionAll` type, this `T` ranges over all
+types, including `Any`, so this type should be equivalent to `Tuple{Any,Any}`.
+However, this interpretation causes some practical problems.
+
+First, a value of `T` needs to be available inside the method definition.
+For a call like `f(1, 1.0)`, it's not clear what `T` should be.
+It could be `Union{Int,Float64}`, or perhaps `Real`.
+Intuitively, we expect the declaration `x::T` to mean `T === typeof(x)`.
+To make sure that invariant holds, we need `typeof(x) === typeof(y) === T` in this method.
+That implies the method should only be called for arguments of the exact same type.
+
+It turns out that being able to dispatch on whether two values have the same type
+is very useful (this is used by the promotion system for example), so we have
+multiple reasons to want a different interpretation of `Tuple{T,T} where T`.
+To make this work we add the following rule to subtyping: if a variable occurs
+more than once in covariant position, it is restricted to ranging over only concrete
+types.
+("Covariant position" means that only `Tuple` and `Union` types occur between an
+occurrence of a variable and the `UnionAll` type that introduces it.)
+Such variables are called "diagonal variables" or "concrete variables".
+
+So for example, `Tuple{T,T} where T` can be seen as
+`Union{Tuple{Int8,Int8}, Tuple{Int16,Int16}, ...}`, where `T` ranges over all
+concrete types.
+This gives rise to some interesting subtyping results.
+For example `Tuple{Real,Real}` is not a subtype of `Tuple{T,T} where T`, because
+it includes some types like `Tuple{Int8,Int16}` where the two elements have
+different types.
+`Tuple{Real,Real}` and `Tuple{T,T} where T` have the non-trivial intersection
+`Tuple{T,T} where T<:Real`.
+However, `Tuple{Real}` *is* a subtype of `Tuple{T} where T`, because in that case
+`T` occurs only once and so is not diagonal.
+
+Next consider a signature like the following:
+
+```julia
+f(a::Array{T}, x::T, y::T) where T = ...
+```
+
+In this case, `T` occurs in invariant position inside `Array{T}`.
+That means whatever type of array is passed unambiguously determines
+the value of `T` --- we say `T` has an *equality constraint* on it.
+Therefore in this case the diagonal rule is not really necessary, since
+the array determines `T` and we can then allow `x` and `y` to be of
+any subtypes of `T`.
+So variables that occur in invariant position are never considered diagonal.
+This choice of behavior is slightly controversial, but your humble documentation
+author (Jeff in this case) likes it.
```

Comment:
I think the triangular rule is referred to as controversial here because I was asking about it, due to the observance of bugs in this explanation of it (also because it is not currently implemented completely, as mentioned later).

Jarret noticed a nice symmetry that appears to let you rewrite covariant constraints as invariant constraints. To do that, the behavior described here seems to require the introduction of O(n) new TypeVars. Keeping the older v0.5 behavior of this instead should only require the introduction of O(1) new TypeVars. The current algorithm (described below @ line 432 https://github.com/JuliaLang/julia/pull/21131/files/a210f10a71f84e2df415494585681034e22491f4#diff-f5c313b6085f51043107224bfc65497eR432) is attempting to provide this O(n) behavior with O(1) new TypeVars. This is not conclusive (i.e. there may actually be a O(1) algorithm), but it is suggestive of the fix I believe may need to be made to correct this class of bug if we keep the new behavior.